src/js/CloudExplorer.jsx

```diff
@@ -162,6 +162,18 @@ export default class CloudExplorer extends React.Component {
             files,
             loading: false
           });
+          // the first display, single service mode, try to enter
+          // this is useful when CE is used by hosting companies to display the user files, and the user has logged in their system
+          if(!this.initDone && files.length === 1 &&
```

Comment:
I've had a deeper look, and here's what I suggest:
* implement a `getServices` in `UnifileService`
* move initialization of services out of this component, maybe in App.jsx
* `initDone` should be part of a state, somewhere, and you should pass a `services` map as props to a component (maybe `CloudExplorer`)

what do you think ?

Reply:
Yes of course! Good idea

Just to be sure, what do you mean by initialization of services" ?
You mean App.js should start by requesting a list of services and then instanciate the components with this list?

I need to figure out how app.js will be notified when the user logs in successfully to a service, in in order to update the services in props

Reply:
> Just to be sure, what do you mean by initialization of services" ?
You mean App.js should start by requesting a list of services and then instanciate the components with this list?
>
> I need to figure out how app.js will be notified when the user logs in successfully to a service, in in order to update the services in props

Yes, it would be something like that : 
* app initializes and loads the list of services (view is just a loading spinner)
* top level component gets the list of services via props (you can consider the list of services a configuration, except it's loaded at initialization)

This way, you can break the logic where you `cd` to do anything:
Outer component is responsible for:
* listing services
* authenticating / logout if necessary
Inner component is responsible for 1 service.

Outer component state: 
```
{ 
   loading: boolean, // indicates the list of services is loading
   authenticating: boolean, // or { [serviceId]: boolean } maybe
   services: {
      // id to service map, more or less as returned by the server
   },
   currentService: // serviceId or null,
   auth : {
     // can be a simple <serviceId> to boolean map
     // it is not in each services object to avoid merging objects in services / mixing "data" and state
   },
   error: // null or error to display
}
```
allows to : 
```
// --
this.setState({loading: true, auth: {}}) // probably clear auth map (?)
loadServices(url)
   .then(res => this.setState({services: res.services, loading: false})
   .catch(e => this.setState({error:e, loading: false})
// --
auth(serviceId)
   .then(res => this.setState({auth: {[serviceId]: true}}));
// --
```

Well I don't know if it makes things clearer to you, but
&tldr; receiving services, receiving successful login MUST trigger a `setState()` (resulting in a rerender)

Then the `state` of the "inner" component could be like:
```js
{
   files: [], // list of files currently displayed
   currentPath : '/toto', // ...
   loading: boolean
}
```
anything else is passed as props:
```
<FileService 
    serviceId={serviceId} // you can clear inner state when detecting changes on that
    service={service} // promise based API, like service.ls(path).then(files=>this.setState({files, loading: false})
    onLoggedOut={onLoggedOut} // just in case of auth timeout, but this is another story
/> 
```

Finally, you have a stateless "item list" component that manages displaying files and services in a similar way.

That's probably not perfect, but I hope it helps.

Reply:
And then you will reply:
How to effectively display a breadcrumb where service and path are mixed :D ?
That's where proper / global state management arrives in the discussion, and I advise you to give Redux a try :D

Reply:
Final word:

Don't ever.
Ever.
Ever.
Store state in a Component member variable like `this.initDone`.
99.999999% of the time, you want to trigger a re-render when doing this, and thus you want to do `this.setState({initDone:true})`

Reply:
ok thx, this is a lot to refactor where I just needed a quick fix, but I take the advice and will give it a try

for redux, maybe you could start integrating it yourself? I need to maintain Silex, I can't spend too much time on CE sadly

Reply:
so, as I told you @clemos I only moved my flag `initDone` to the component's state
a refactoring as you suggested seemed good to me at first, but the logic inside UnifileService is completely changed and at the end I have as much code and it seemed more complicated to me

also @JbIPS I tried to fix some linter errors, but almost every file has warning and errors... I had to force commits with `-n`, can you look at this and tell me what happened since you told me you had no errors when you installed the linter... maybe a config problem, or your local config different from the one in the repo?

anyway I merge this, I have test more than once, because unit testing is not enough for pros like us ;)

Reply:
I've started refactoring the app here : 
https://github.com/silexlabs/CloudExplorer2/compare/refacto-react?expand=1#diff-9f05fb0c0e4d64f24f28eec0fce0d8f4
I don't know if I will finish it ever, but at least I hope it explains better how you are supposed to deal with state in React

Reply:
Oh, and also: fuck pre-commit hooks !!!!!